File: DSA/LinkedList/java/_1.java
package DSA.LinkedList.java;

public class _1 {
    public static void main(String[] args) {
        Node ll = new Node(10);
        Node temp1 = new Node(5);
        Node temp2 = new Node(7);
        ll.next = temp1;
        temp1.next = temp2;
        while (ll != null) {
            int x = ll.data;
            System.out.print(x);
            System.out.print(" ");
            ll = ll.next;
        }
        System.out.println();
        ll = insertAtHead(78, ll);
        ll = insertAtHead(12, ll);
        PrintList(ll);
        System.out.println();
        ll = insertAtTail(90, ll);
        ll = insertAtTail(67, ll);
        PrintList(ll);
        System.out.println();
        ll = insertAtHead(13, ll);
        ll = insertAtHead(14, ll);
        ll = insertAtHead(53, ll);
        ll = insertAtHead(63, ll);
        ll = insertAtTail(90, ll);
        ll = insertAtTail(67, ll);
        ll = deleteAtHead(ll);
        PrintList(ll);
        System.out.println();
    }

    public static void PrintList(Node head) {
        if (head == null)
            return;
        System.out.print(head.data + " ");
        PrintList(head.next);
    }

    public static Node insertAtHead(int x, Node head) {
        Node ll = new Node(x);
        ll.next = head;
        return ll;
    }

    public static Node insertAtTail(int x, Node head) {
        Node ll = new Node(x);
        Node temp = head;
        while (temp.next != null) {
            temp = temp.next;
        }
        temp.next = ll;
        return head;
    }

    public static Node deleteAtHead(Node head) {
        if (head == null) {
            return null;
        } else {
            return head.next;
        }
    }

    public static Node deleteAtTail(Node head) {
        if (head == null)
            return null;
        if (head.next == null)
            return null;
        Node temp = head;
        while (temp.next != null)
            temp = temp.next;
        temp.next = null;
        return head;
    }
}

File: DSA/LinkedList/cpp/_1.cpp
#include <iostream>
using namespace std;
struct Node
{
  int data;
  Node *next;
  Node(int x)
  {
    this->data = x;
    this->next = nullptr;
  }
};
Node *insertAtHead(Node *head, int x)
{
  Node *ll = new Node(x);
  ll->next = head;
  return ll;
}
Node *insertAtTail(Node *head, int x)
{
  Node *ll = new Node(x);
  Node *temp = head;
  while (temp->next != nullptr)
  {
    temp = temp->next;
  }
  temp->next = ll;
  return head;
}
Node *deleteAtHead(Node *head)
{
  if (head == NULL)
    return NULL;
  else
  {
    // return head->next;
    Node *temp = head->next;
    delete head;
    return temp;
  }
}
Node *deleteAtTail(Node *head)
{
  if (head == nullptr)
    return nullptr;
  if (head->next == nullptr)
  {
    delete head;
    return nullptr;
  }
  Node *temp = head;
  while (temp->next->next != nullptr)
  {
    temp = temp->next;
  }
  delete temp->next;
  temp->next = nullptr;
  return head;
}
void PrintList(Node *head)
{
  if (head == nullptr)
    return;
  cout << head->data << " ";
  PrintList(head->next);
}
int main()
{
  Node *ll = new Node(5);
  ll->next = new Node(10);
  ll->next->next = new Node(15);
  while (ll)
  {
    int x = ll->data;
    cout << x << " ";
    ll = ll->next;
  }
  cout << "\n";
  Node *head = new Node(10);
  Node *temp1 = new Node(29);
  Node *temp2 = new Node(54);
  head->next = temp1;
  temp1->next = temp2;
  while (head)
  {
    int x = head->data;
    cout << x << " ";
    head = head->next;
  }
  cout << "\n";
  head = insertAtHead(head, 90);
  head = insertAtHead(head, 98);
  PrintList(head);
  cout << "\n";
  head = insertAtTail(head, 89);
  head = insertAtTail(head, 99);
  head = insertAtTail(head, 809);
  head = insertAtTail(head, 898);
  head = insertAtTail(head, 891);
  head = insertAtTail(head, 189);
  head=deleteAtHead(head);
  head=deleteAtTail(head);
  head=deleteAtTail(head);
  PrintList(head);
  cout << "\n";
  return 0;
}